pkg/storageos/scheduler_extender.go

```diff
@@ -0,0 +1,280 @@
+		},
+	}
+}
+
+// schedulerPolicyTemplate contains fields for rendering the scheduler policy.
+type schedulerPolicyTemplate struct {
+	FilterVerb     string
+	PrioritizeVerb string
+	EnableHTTPS    bool
+	URLPrefix      string
+}
+
+// createSchedulerPolicy creates a configmap with kube-scheduler policy.
+func (s Deployment) createSchedulerPolicy() error {
+	policyConfigMap := s.getConfigMap(policyConfigMapName)
+	policyConfigurationTemplate := `
+    {
+      "kind" : "Policy",
+      "apiVersion" : "v1",
+      "predicates" : [
+        {"name" : "PodFitsHostPorts"},
+        {"name" : "PodFitsResources"},
+        {"name" : "NoDiskConflict"},
+        {"name" : "MatchNodeSelector"},
+        {"name" : "HostName"}
+      ],
+      "extenders" : [{
+        "urlPrefix": "{{.URLPrefix}}",
+        "filterVerb": "{{.FilterVerb}}",
+        "prioritizeVerb": "{{.PrioritizeVerb}}",
+        "weight": 1,
+        "enableHttps": {{.EnableHTTPS}},
+        "nodeCacheCapable": false
+      }]
+    }
+`
+	// Service address format: <service-name>.<namespace>.svc.cluster.local.
+	serviceEndpoint := fmt.Sprintf("%s.%s.svc.cluster.local", s.stos.Spec.GetServiceName(), s.stos.Spec.GetResourceNS())
+	policyData := schedulerPolicyTemplate{
+		FilterVerb:     "filter",
+		PrioritizeVerb: "prioritize",
+		EnableHTTPS:    false,
+		URLPrefix:      fmt.Sprintf("http://%s:5705/v1/scheduler", serviceEndpoint),
+	}
+
+	// Render the policy configuration.
+	var policyConfig bytes.Buffer
+	tmpl, err := template.New("policyConfig").Parse(policyConfigurationTemplate)
+	if err != nil {
+		return err
+	}
+	if err := tmpl.Execute(&policyConfig, policyData); err != nil {
+		return err
+	}
+
+	// Add the policy configuration in the configmap.
+	policyConfigMap.Data = map[string]string{
+		"policy.cfg": policyConfig.String(),
+	}
+	return s.createOrUpdateObject(policyConfigMap)
+}
+
+// schedulerConfigTemplate contains fields for rendering the scheduler
+// configuration.
+type schedulerConfigTemplate struct {
+	SchedulerName  string
+	PolicyName     string
+	Namespace      string
+	LeaderElection bool
+}
+
+// createSchedulerConfiguration creates a configmap with kube-scheduler
+// configuration.
+func (s Deployment) createSchedulerConfiguration() error {
+	configConfigMap := s.getConfigMap(schedulerConfigConfigMapName)
+	configTemplate := `
+    apiVersion: "kubescheduler.config.k8s.io/v1alpha1"
+    kind: KubeSchedulerConfiguration
+    schedulerName: {{.SchedulerName}}
+    algorithmSource:
+      policy:
+        configMap:
+          namespace: {{.Namespace}}
+          name: {{.PolicyName}}
+    leaderElection:
+      leaderElect: {{.LeaderElection}}
+      lockObjectName: {{.SchedulerName}}
+      lockObjectNamespace: {{.Namespace}}
+`
+	schedConfigData := schedulerConfigTemplate{
+		SchedulerName:  schedulerExtenderName,
+		PolicyName:     policyConfigMapName,
+		Namespace:      s.stos.Spec.GetResourceNS(),
+		LeaderElection: true,
+	}
```

Comment:
Curious how the leader election works?  I saw you've set replicas to 1 - do we need to test > 1?

Reply:
I think we should provide an option in the cluster config to set this number, defaulting to 1 or 2. I've not looked into leader election. We do need to test it. Also, maybe we should add pod toleration for quick recovery, similar to the ones we have for CSI helper deployment pods.